No more ComputeStage

lindsayad opened the pull request:
I'm just throwing this up for testing right now. On a incompressible Navier Stokes (INS) problem with Sparse AD, the run time only increases from 4.7 minutes to 4.85 minutes when creating every AD object as  a `JACOBIAN` object and just not seeding during the residual computation phase.

I'm guessing we're going to get a bunch of timeouts here since the default MOOSE configuration is with non-sparse AD.

I'm also going to do some tests on other problems since INS demands an expensive preconditioning process.

lindsayad committed:
Remove ComputeStage from the framework

Removing ComputeStage is representative of the fact that we are
no longer going to have duplicated AD objects for residual and
Jacobian calculation

Refs #14547

dschwen commented:
How is this dealing with external library calls that require `Real` to be `double`. I.e. FParser and all the BLAS/LAPACK stuff. I know we have some specialization for BLAS/LAPACK stuff, but they are not feature complete and have known issues.

lindsayad commented:
Perhaps the title is a little misleading...`Real` still means the tradiitonal POD. It's just that we would remove `ComputeStage` so there would be no more need to two copies of AD objects. Instead we use sparse AD and just don't see during residual computation.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/465255/) on fc5ae0d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14701/site)

This comment will be updated on new commits.

lindsayad commented:
so I'm looking at test timeouts from [here](https://civet.inl.gov/job/465247/). There are 14. 11 of them are PETSc Jacobian tests. It doesn't surprise me that we would get timeouts on PETSc Jacobian tests in debug mode because we do a fair amount of assertions down in MetaPhysicL and we are making a billion DualNumber calls during Jacobian testing because of finite differencing of the residuals. So looking more seriously at the other timeouts:

### rom_stress_update

In oprof mode with a `uniform_refine=1`:
- all AD: 15.8 seconds, residual is 15.6% of timing
- current: 11.8 seconds, residual is 4.2% of timing

aeslaughter commented:
Do you envision being able to continue to optimize the functionality or is it as good as it will be?

What is the use case for PJFNK with AD?

lindsayad commented:
> Do you envision being able to continue to optimize the functionality or is it as good as it will be?

The heaviest function does one integer operation (`operator-`) and one pointer operation. It's possible that @rwcarlsen's boolean flag would reduce some overhead by immediately returning for `DualNumber` unary and binary operations as opposed to going down to `DynamicSparseNumberBase:;sparsity_union` and checking whether the current iterator matches the `end` iterator.

> What is the use case for PJFNK with AD?

There are some cases, particularly for thermal and mechanical contact problems, where for the current state of AD the Jacobian is not perfect. This  could be alleviated when geometric AD is in libMesh.

fdkong commented:
> @idaholab/moose-team Removing `ComputeStage` and controlling AD cost by simply not seeding during residual evaluation will be fairly costly if the user is doing PJFNK. If the user is using NEWTON, the cost is fairly small. So how do we want to proceed from here?

At least we should set one flag in dual number (just as @rwcarlsen  did before) for residual calculations so that derivative calculations are just skipped.

And then,  I believe we will have similar performances.

joshuahansel commented:
I don't understand the mechanics of AD well, but where is the additional cost incurred with this change? As I understand, AD classes are templated on `ComputeStage`, which is `Residual` or `Jacobian`, the former which does not do any derivative computations. And now they are all `Jacobian`, but there is no seeding during residual evaluation. So is additional cost incurred by the sparse AD part or the all-Jacobian-stage part?

Regarding PJFNK vs. Newton, are there also cases where even with a perfect Newton Jacobian, you'd choose PJFNK?

dschwen commented:
Ok, so this is basically simplifying the ADKernel system and its underpinnings.

lindsayad commented:
> I don't understand the mechanics of AD well, but where is the additional cost incurred with this change? As I understand, AD classes are templated on `ComputeStage`, which is `Residual` or `Jacobian`, the former which does not do any derivative computations. And now they are all `Jacobian`, but there is no seeding during residual evaluation. So is additional cost incurred by the sparse AD part or the all-Jacobian-stage part?

You can check out the performance graphs. The cost is in the sparse AD part, although the cost would be much worse with nonsparse AD (using MetaPhysicL master). As I mentioned in https://github.com/idaholab/moose/pull/14701#issuecomment-587210548 the heaviest function is simply checking the `end()` of the sparse derivative container which involves one integer subtraction and then one pointer arithmetic operation (`+`). Obviously this is a very simple function, so its cost is simply because its called a billion times. I tried saving adding an `_offset` member to the MetaPhysicL class to avoid doing the integer operation, but then you just have to do another assignment when copy constructing the derivative container, so it seems that there is no gain. @rwcarlsen told me that when did a `boolean` flag in the `DualNumber` class to avoid derivative computation that he was able to avoid any penalty, so I'm holding out some hope.

> Regarding PJFNK vs. Newton, are there also cases where even with a perfect Newton Jacobian, you'd choose PJFNK?

If memory resources are a concern. I think @friedmud says the crossover point is around 10 variables or something.

lindsayad commented:
> Ok, so this is basically simplifying the ADKernel system and its underpinnings.

Exactly correct. It's for improving application developer (and even MOOSE developer) experience. At least one thing this has shown me is that the original design was a good one in terms of performance :smile:

lindsayad committed:
Toggle MetaPhysicL flag in FEProblemBase

lindsayad commented:
Ok, if we add a `static bool do_derivatives` data member like in roystgnr/MetaPhysicL#56, the computation time for the `powerlaw_sum` goes to:

- Non-sparse: 6.84 seconds [non-sparse-profile](https://user-images.githubusercontent.com/10248304/74976039-dfbf7b00-53dc-11ea-8fff-2dc86339d258.png)
    -  previously 55 seconds
- Sparse: 6.28 seconds [sparse-profile](https://user-images.githubusercontent.com/10248304/74975889-9f5ffd00-53dc-11ea-938a-eea40e3f70e3.png)
    - previously 13 seconds

rwcarlsen commented:
You mean the static bool set to true is 2x slower than the templated case with no if-check for your powerlaw_sum benchmark? - which is much faster than just the no-derivative-seeding slowdowns? (your comment seems a bit unclear to me)  If so than that seems like good enough to me.  Especially when you consider how much it affects overall larger simulation times - which should be tiny.  Have you rerun any of the navier stokes cases with the bool checking?

lindsayad commented:
- case 1: `static bool` set to `true` during Jacobian evaluation and to `false` during residual evaluation
- case 2: current templated case
- case 3: sparse, don't seed during residual evaluations, no changes to `MetaPhysicL`
- case 4: non-sparse, don't seed during residual evaluations (this doesn't matter for non-sparse), no changes to `MetaPhysicL`

Results summary for the `powerlaw_sum` benchmark:
- case 1 is 2-3x slower than case 2
- case 1 is 2x faster than case 3
- case 1 is almost 10x faster than case 4

>  Especially when you consider how much it affects overall larger simulation times - which should be tiny.

If a user is doing AD with PJFNK I would expect proportional slow-downs regardless of the scale of the simulation because they're going to be doing `DualNumber` temporary construction for every element computation.

>  Have you rerun any of the navier stokes cases with the bool checking?

I have not. NEWTON simulations, especially with expensive preconditioners, should be fine for us because we do so much fewer residual evaluations

fdkong commented:
> Ok, if we add a `static bool do_derivatives` data member like in [roystgnr/MetaPhysicL#56](https://github.com/roystgnr/MetaPhysicL/pull/56), the computation time for the `powerlaw_sum` test case above goes to:
> 
>     * Non-sparse: 6.84 seconds [non-sparse-profile](https://user-images.githubusercontent.com/10248304/74976039-dfbf7b00-53dc-11ea-8fff-2dc86339d258.png)
>       
>       * previously 55 seconds
> 
>     * Sparse: 6.28 seconds [sparse-profile](https://user-images.githubusercontent.com/10248304/74975889-9f5ffd00-53dc-11ea-938a-eea40e3f70e3.png)
>       
>       * previously 13 seconds
> 
> 
> This is now 2-3x slow-down compared to current MOOSE which runs at 2.7 seconds

This is about one residual evaluation?

rwcarlsen commented:
The penalty still will be amortized over some level of "fixed" cost of nonlinear and linear iterations.  Real simulation slowdown I bet would be more like 1.5X or less even depending on the simulation.

permcody commented:
@lindsayad - There's a cool new feature on Github. You can now open up 'draft PRs'. These cannot be merged until marked by the author as ready. You might try that next time you have a work in progress.

lindsayad committed:
Update to new DualNumber template signature

lindsayad committed:
Use registerMooseObject in framework

lindsayad committed:
ADMaterialPropertyObject -> ADMaterialProperty

lindsayad committed:
Purge *Templs and associated macros

Refs #14547

lindsayad committed:
Don't treat AD and regular objects differently

Remove discrimination between AD and regular objects from the
Registry, FEProblemBase add* methods, and warehouse storage

Refs #14547

lindsayad committed:
Call computeADOffDiagJacobian through computeOffDiagJacobian

Add a boolean flag such that computeADOffDiagJacobian only gets
called the first time computeOffDiagJacobian gets called (since
the former method computes all the derivatives for all the coupled
variables at once).

Refs #14547

lindsayad committed:
Remove ComputeStage from test

Refs #14547

lindsayad committed:
registerADMooseObject -> registerMooseObject everywhere else

lindsayad committed:
Notify *Problem when we have AD objects

Refs #14547

lindsayad committed:
Purge ComputeStage from  modules

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/490275/) on 5278e78 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14701/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14701/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5a6fabeba1e6a3bee85911f1bb3e46bba38dfc42`

lindsayad commented:
@idaholab/moose-team The currently failing phase field tests raise the quandary: Do we want to continue this inter-operability of AD and non-AD material properties? The problem now is this:

@rwcarlsen's refactoring of `MooseADWrapper` was excellent, but even with the vastly improved design, we are still open to loopholes which is what was proved by the phase field tests: they simply initialize their material property during `initialSetup` and never touch it again. In the old days of AD this worked fine: the first time we ever called `MooseADWrapper:dn` was during Jacobian calculation and we initialized the derivative vector and everything was hunky-dory. But now, the first time we call `MooseADWrapper::dn` is during *residual* computation, and then we *don't* initialize the derivative vector. We then mark `_dual_is_dirty = false`, and then when we actually call `MooseADWrapper::dn` during *Jacobian* evaluation, we don't do anything and the derivative vector stays uninitalized; this causes all kinds of valgrind errors and undefined Jacobian entries when we then actually try to use the derivative vector. The fix for this phase field case would be fairly straightforward but also waste a bunch of computation: actually implement a `computeQpProperties` method that does stupid/unnecessary filling of the material property; this would mark `_dual_is_dirty` in `MooseADWrapper` and then synchronization would happen during the Jacobian evaluation. But what's going to stop an unwary developer from doing something like this in the future, especially when naively it's a very smart thing for them to do!

Another fix would be: get rid of `MooseADWrapper` and just have every `Real` in every `MaterialProperty` be a `DualReal`. Then there would never be any synchronization concerns. ***But*** then we are vastly increasing memory requirements, particularly in the case of stateful material properties.

A final fix, and perhaps the one that I am most inclined to adopt, is to get rid of the interoperability of non-AD and AD material properties. The most annoying thing I can think of here is that you couldn't use stuff like `GenericConstantMaterial` to fill AD material properties. Another annoyance is that for non-sparse AD, during Jacobian evaluation you'd be be doing some extra computation for truly constant material properties if they're necessarily required to be AD. But I don't think these expense concerns outweigh the guaranteed safety/correctness from going with this concept. (This concept also allows the elimination of `MooseADWrapper`)

joshuahansel commented:
@lindsayad To clarify: when you say that you propose getting rid of "inter-operability of AD and non-AD material properties", you mean that if you have one AD material property, then all material properties must be AD? If this is the case, then my only note will be that this could make it a bit of an ordeal for non-AD applications to transition to AD: it'd mean an all-in-one-go shot for switching over material properties (and all of the objects that retrieve them, right?).

If that's the case, but it's the right thing for the framework in the long run, then I think it should be done.

lindsayad commented:
> @lindsayad To clarify: when you say that you propose getting rid of "inter-operability of AD and non-AD material properties", you mean that if you have one AD material property, then all material properties must be AD? If this is the case, then my only note will be that this could make it a bit of an ordeal for non-AD applications to transition to AD: it'd mean an all-in-one-go shot for switching over material properties (and all of the objects that retrieve them, right?).

No, it wouldn't be nearly that restrictive. Basically what it would mean is that you can only ever use these pairs: `declareProperty, getMaterialProperty` and `declareADProperty, getADMaterialProperty`. Currently you can mix and match like: `declareADProperty,getMaterialProperty` or `declareProperty,getADMaterialProperty`

joshuahansel commented:
> No, it wouldn't be nearly that restrictive. Basically what it would mean is that you can only ever use these pairs: declareProperty, getMaterialProperty and declareADProperty, getADMaterialProperty. Currently you can mix and match like: declareADProperty,getMaterialProperty or declareProperty,getADMaterialProperty

So as for transitioning, I could transition a material property to AD, but then I'd need to change to `getADMaterialProperty` in all of the objects using it, and thus I'd need to change them to AD as well at that point?

lindsayad commented:
> So as for transitioning, I could transition a material property to AD, but then I'd need to change to `getADMaterialProperty` in all of the objects using it, and thus I'd need to change them to AD as well at that point?

If we go the route I'm touting, then yes everything you say here is correct.

joshuahansel commented:
On second thought, I don't think this would actually be that bad - I'd just need to start at the Kernel-end and end at the Material-end. That is, I can gradually move over Kernels and such, and then when those are done, then I can gradually move over Materials.

dschwen commented:
> just have every `Real` in every (AD)MaterialProperty be a `DualReal`

That's a no go as far as inelastic deformation mechanics go.

fdkong commented:
> No, it wouldn't be nearly that restrictive. Basically what it would mean is that you can only ever use these pairs: `declareProperty, getMaterialProperty` and `declareADProperty, getADMaterialProperty`. Currently you can mix and match like: `declareADProperty,getMaterialProperty` or `declareProperty,getADMaterialProperty`

From the framework point of view, I would like the idea (without mixing):

 `declareProperty, getMaterialProperty` and `declareADProperty, getADMaterialProperty`


Mixing together has a lot of maintenance headaches. Potentially creates a lot of corner cases.

joshuahansel commented:
I'd also add that `declareProperty`, `getADMaterialProperty` is potentially risky anyway because someone could grab a material property, not realizing it's not AD, and thus be missing derivatives without their knowledge.

lindsayad committed:
Remove interoperability of regular and AD material properties

Refs #14547

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/492311/) on a5825b2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14701/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14701/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5a6fabeba1e6a3bee85911f1bb3e46bba38dfc42`

lindsayad committed:
Demonstrate generic Material/ADMaterial

Refs #14547

lindsayad commented:
@dschwen See the changes to `GenericConstantMaterial` in https://github.com/idaholab/moose/pull/14701/commits/4a393984f22ad9f63fdaed01d215380aac099a94 for how I think a user could limit code duplication between a "regular" and an "AD" material

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/492396/) on 4a39398 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14701/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14701/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8fdf5f581b7862a553b35a53d221328f8332d633`

lindsayad committed:
Be able to output AD material properties in MaterialOutputAction

Refs #14547

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit ef7dd34ad117998e7b7ef909e85a6ae53f57e057
+Subproject commit 179e8c8c2dcbfcc01311bd6f5044a9534aa21584
```
**Caution!** This contains a submodule update

lindsayad committed:
Update modules for removal of AD/non-AD material prop interoperability

Refs #14547

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/493458/) on fc6a771 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14701/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14701/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 277450edfdecd33544d89cc7ab224043565fa096`

lindsayad committed:
TM tests: we no longer support mixing non-AD/AD declare/get

lindsayad committed:
Update TensorMechanics

Update TensorMechanics for removal of regular/AD material property
interoperability. Involved:

- Create AD versions of constant materials, auxes, postprocessor
  using a boolean is_ad template parameter to limit code duplication
- Adjust relevant input files to use new AD objects

Refs #14547

lindsayad committed:
Update phase_field

Refs #14547

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/493973/) on d629498 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14701/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14701/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6cce03c6eb7b9914973170ebd54fdb730f8149e4`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/494352/) on 7f660a2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14701/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14701/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 55a2368dca45913ae0385c4d5fcff68318d3e4ef`

lindsayad committed:
Update contact

lindsayad committed:
Update heat_conduction

Refs #14547

lindsayad committed:
Update navier_stokes

Refs #14547

lindsayad committed:
Update combined

Refs #14547

lindsayad committed:
Remove thermal_elastic/ad-derivatives test

This test did nothing for verifying the accuracy of the Jacobian
coming out of ADComputeVariableIsotropicElasticityTensor. This
passed Jacobian testing because the test kernel computed all the
derivatives through DerivativeParsedMaterial, not through
MetaPhysicL AD. The truth of the matter is that DerivativeParsedMaterial
and MetaPhysicL AD are not setup yet to communicate with each other
(although they should be soon)

Refs #14547

lindsayad commented:
Now I sit and wait for idaholab/conda-moose#71 to go in before this can be really tested

milljm commented:
@lindsayad It's in. Though I messed up a little bit. You _shouldn't_ run into any issues from my mistake (I updated the 'moose' package which no one uses, without bumping the build number). I'll fix that later today.

This PR should be ready to be tested/invalidated to use latest libMesh (2020.03.31).

lindsayad committed:
Add equality helper for vector material properties

Refs #14547

lindsayad committed:
Update unit

Refs #14547

lindsayad committed:
Update porous_flow

Refs #14547

lindsayad committed:
Check do_derivatives flag before reading derivative info

Refs #14547

lindsayad committed:
Update tutorial

Refs #14547

lindsayad committed:
Initialize DualReal::do_derivatives early

Initialize to true so that when creating default AD values early
in the simulation, the derivative vectors actually get initialized.
Otherwise we're facing a ticking time bomb of reading from an
uninitialized value

Refs #14547

lindsayad commented:
Oh wow, all MOOSE test targets passing...

lindsayad committed:
Work for Apps

Refs #14547

lindsayad committed:
Add MaterialConverter class

This class can be used for explicit conversions between
regular and AD material properties

Refs #14547

lindsayad commented:
@idaholab/moose-team I've finally gotten all (default) BISON tests passing, so I think that this is ready for review

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/498923/) on 40423d8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14701/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14701/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 923ae514285f9b88165a9cbeda17cc51562fcce4`

lindsayad committed:
clang-format v8

rwcarlsen commented:
Looks like a test spec file problem:

```
Parser Errors:
--------------------------------------------------------------------------------------------------------------
/opt/civet/build_0/moose/modules/tensor_mechanics/test/tests/ad_finite_strain_jacobian/tests:85: duplicate parameter "/Tests/3d_bar_Bbar-jac/cli_args"
--------------------------------------------------------------------------------------------------------------
```

lindsayad committed:
Restrict thread count for stateful mortar adaptivity test

rwcarlsen commented:
For purposes of review, could you suggest some places/changes that perhaps warrant more scrutiny than just the mundane removing templating boiler-plate changes?  Just looking for a bit of a map to make my review as effective as possible.

lindsayad commented:
> For purposes of review, could you suggest some places/changes that perhaps warrant more scrutiny than just the mundane removing templating boiler-plate changes? Just looking for a bit of a map to make my review as effective as possible.

Probably `MooseADWrapper, MaterialProperty, MaterialPropertyStorage::shift` are the most important changes I can think of off the top of my head. Other than that it's erasing the template, removing the macros that were necessitated by the templating, and toggling the `do_derivatives` to `false` when doing residual evaluations.

tophmatthews commented:
> BISON is the app that most heavily uses AD

You're welcome :) There'll hopefully more after these huge reworking making things a lot easier!

lindsayad committed:
Disable valgrind testing for large navier stokes supg mms tests

Valgrind tests are already conducted on smaller mesh problems
that otherwise have identical inputs so we do not have to worry
about losing different code path testing

rwcarlsen commented on framework/include/materials/MaterialBase.h:
```diff
@@ -101,11 +101,23 @@ class MaterialBase : public MooseObject,
+  MaterialProperty<T> & declarePropertyOlder(const std::string & prop_name);
+  template <typename T>
+  ADMaterialProperty<T> & declareADProperty(const std::string & prop_name);
+  template <typename T, bool is_ad, typename std::enable_if<is_ad, int>::type = 0>
+  ADMaterialProperty<T> & declareGenericProperty(const std::string & prop_name)
+  {
+    return declareADProperty<T>(prop_name);
+  }
+  template <typename T, bool is_ad, typename std::enable_if<!is_ad, int>::type = 0>
+  MaterialProperty<T> & declareGenericProperty(const std::string & prop_name)
+  {
+    return declareProperty<T>(prop_name);
+  }
```
Could you help me understand what role this function and the is_ad toggling is serving?

lindsayad replied:
It allows creation of generic `Material` templates that can either be selected to be regular or AD. This is my example:
```c++
registerMooseObject("MooseApp", GenericConstantMaterial);
registerMooseObject("MooseApp", ADGenericConstantMaterial);

template <bool is_ad>
InputParameters
GenericConstantMaterialTempl<is_ad>::validParams()
{
  InputParameters params = Material::validParams();
  params.addRequiredParam<std::vector<std::string>>(
      "prop_names", "The names of the properties this material will have");
  params.addRequiredParam<std::vector<Real>>("prop_values",
                                             "The values associated with the named properties");
  params.declareControllable("prop_values");
  return params;
}

template <bool is_ad>
GenericConstantMaterialTempl<is_ad>::GenericConstantMaterialTempl(
    const InputParameters & parameters)
  : Material(parameters),
    _prop_names(getParam<std::vector<std::string>>("prop_names")),
    _prop_values(getParam<std::vector<Real>>("prop_values"))
{
  unsigned int num_names = _prop_names.size();
  unsigned int num_values = _prop_values.size();

  if (num_names != num_values)
    mooseError(
        "Number of prop_names must match the number of prop_values for a GenericConstantMaterial!");

  _num_props = num_names;

  _properties.resize(num_names);

  for (unsigned int i = 0; i < _num_props; i++)
    _properties[i] = &declareGenericProperty<Real, is_ad>(_prop_names[i]);
}

template <bool is_ad>
void
GenericConstantMaterialTempl<is_ad>::initQpStatefulProperties()
{
  computeQpProperties();
}

template <bool is_ad>
void
GenericConstantMaterialTempl<is_ad>::computeQpProperties()
{
  for (unsigned int i = 0; i < _num_props; i++)
    (*_properties[i])[_qp] = _prop_values[i];
}

template class GenericConstantMaterialTempl<false>;
template class GenericConstantMaterialTempl<true>;
```

rwcarlsen commented on framework/include/materials/MaterialData.h:
```diff
@@ -296,19 +358,42 @@ MaterialData::getProperty(const std::string & name)
+    // We didn't find a regular material property so we're going to error out. But we can check to
+    // see whether there is an AD property of the same name in the hope that we can give the user a
+    // more meaningful error message
+    auto ad_prop = dynamic_cast<ADMaterialProperty<T> *>(_props[prop_id]);
+    if (ad_prop)
+      mooseError("The requested regular material property " + name +
+                 " is declared as an AD property. Either retrieve it as an AD property with "
+                 "getADMaterialProperty or declare it as a regular property wtih declareProperty");
+    else
+      mooseError("Material has no property named: " + name);
```
What was the motivation behind removing graceful fallback from AD to normal properties?  It was definitely useful - especially when incrementally converting apps over to AD.

It's really too bad our material API doesn't have users calling functions at the use-point - then it would be trivial to gracefully throw away derivatives to provide a regular property value when an AD one was defined. Or throw warnings, etc.

lindsayad replied:
I think this comment is the most forefront illustration: https://github.com/idaholab/moose/pull/14701#issuecomment-605265668.

I also like @joshuahansel's comment here: https://github.com/idaholab/moose/pull/14701#issuecomment-606153585

rwcarlsen replied:
Yeah - the interoperability was very confusing and multiple times was the source of major problems - probably needed to go.  It's really too bad we don't have function calls to tap into - this and so many other problems would just be complete non-issues.

lindsayad replied:
That's very true. If a user wants to explicitly throw away things (or even try to add derivatives?) they should be able to. Perhaps that's something we can add later

rwcarlsen commented on framework/include/materials/MaterialProperty.h:
```diff
@@ -170,123 +178,193 @@ class MaterialProperty : public PropertyValue
-template <typename T>
-inline void
-MaterialProperty<T>::markAD(bool use_ad)
+namespace MaterialPropertyPrivate
```
I've been trying to use an `moose::internal` namespace nesting for this kind of stuff.

rwcarlsen reviewed (changes requested):
Seems mostly reasonable - I mostly looked at the materials stuff.  I just had a few minor questions/comments.

rwcarlsen commented:
I like the commit history hygiene.

fdkong commented on framework/src/kernels/ADDiffusion.C:
```diff
@@ -9,31 +9,21 @@
-template <ComputeStage compute_stage>
 ADRealVectorValue
-ADDiffusion<compute_stage>::precomputeQpResidual()
+ADDiffusion::precomputeQpResidual()
```
Finally understand a bit about diffusion equation

fdkong reviewed (dismissed):
The interface definitely looks much better now. General questions:

1. Do we always need to create two versions of kernel for the same physics? AD and nonAD

2. It looks like we are able to able to create a single version of material, and then use it for AD and nonAD using the `<is_ad>` trick. Is it a recommended way to create a new material in the future?

lindsayad commented:
> The interface definitely looks much better now. General questions:
> 
>     1. Do we always need to create two versions of kernel for the same physics? AD and nonAD
> 
If the user wants a hand-coded version then yes they will have to create a second object. `Kernel` and `ADKernel` do fundamentally different things behind the scenes and their API is different

>     2. It looks like we are able to able to create a single version of material, and then use it for AD and nonAD using the `<is_ad>` trick. Is it a recommended way to create a new material in the future?

No, I just recommend creating a single non-templated material class that declares properties which could be AD using `declareADProperty` and only declares regular material properties for properties that are fundamentally constant, e.g. the charge of an electron, Boltzmann's constant, etc. `GenericConstantMaterial` is a bit of a unique case because its name suggests that its properties should be constant. But an `ADGenericConstantMaterial` needs to exist in case we want to make a property that could in general be a function of the nonlinear dofs constant. E.g. `ADHeatConduction` wants an AD thermal conductivity; but sometimes we may just want to use a constant so we can use `ADGenericConstantMaterial`

lindsayad commented on framework/include/actions/MaterialOutputAction.h:
```diff
@@ -45,6 +45,14 @@ class MaterialOutputAction : public Action
 
+  /*
+   * Helper method for testing if the material exists as a block or boundary material
+   * @tparam T The AD property type (e.g., Real)
```
Yes. http://www.doxygen.nl/manual/commands.html#cmdtparam

lindsayad commented on framework/include/auxkernels/MaterialRealVectorValueAux.h:
```diff
@@ -13,15 +13,19 @@
+template <bool>
+class MaterialRealVectorValueAuxTempl;
+typedef MaterialRealVectorValueAuxTempl<false> MaterialRealVectorValueAux;
+typedef MaterialRealVectorValueAuxTempl<true> ADMaterialRealVectorValueAux;
```
Yes since you cannot mix `getADMaterialProperty` and `getMaterialProperty` we need to different AuxKernel versions depending on which we want to retrieve

lindsayad commented on framework/include/base/Assembly.h:
```diff
@@ -223,10 +219,16 @@ class Assembly
+    _calculate_face_xyz = true;
+
+    _calculate_ad_coord = true;
+    return _ad_coord;
```
What do you mean?

lindsayad commented on framework/include/interfaces/Coupleable.h:
```diff
@@ -828,24 +787,21 @@ class Coupleable
    */
-  template <ComputeStage compute_stage>
-  const ADVariableValue & adZeroValueTemplate();
+  const ADVariableValue & adZeroValue();
```
Yea, and that's what this returns :smile:

lindsayad commented on framework/include/kernels/ADMatDiffusionBase.h:
```diff
@@ -37,34 +32,32 @@ class ADMatDiffusionBase : public ADKernelGrad<compute_stage>
 }
 
-template <ComputeStage compute_stage, typename T>
+template <typename T>
```
`_diffusivity[_qp]` may either be a scalar or a tensor if the diffusivity is anisotropic
